keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,454 @@
+I want to migrate all pods available to user B to new nodes, and I want to ensure HA of my services during the migration.
+
+Without `node-affinity-eviction`, I have to remove "userB=allow" label of node, and delete the pods manually.
+With `node-affinity-eviction`, I can simply delete the "userB=allow" label from the existing nodes 
+to re-schedule all pods of user B to these new nodes.
+
+#### handle data locality pods when data is corrupted
+
+I have one application whose pods need to be co-located with some specific data stored in some particular nodes. 
+When one node's data is corrupted, the pod on this node will provide some inaccurate services.
+So I want it to be scheduled to another special node when data is detected to be corrupted.
+And in the meantime, I want to keep other running pods on that node intact.
+Also, I want to ensure my PDB is guaranteed, so I can't taint the node directly.
+
+[stack overflow link](https://stackoverflow.com/questions/55006205/how-to-prevent-pods-scheduling-to-a-node-after-a-pod-has-failed-on-it-in-gce)
+
+Without `node-affinity-eviction`, I have to relabel the node to suggest data on that node is corrupted, 
+and then manually iterating the pods on the node, decide whether to delete them to trigger a reschedule.
+With `node-affinity-eviction`, the only thing I need to do is to relabel the node, and `node-affinity-eviction` controller will do the rest.
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
+
+The major concern for this controller may be the overhead of the nodes with a large number of pods. 
+Every time a node's label is updated, the controller needs to iterate all pods assigned to the node. This may cause potential scaling issues.
+However, the impact won't be significant due to the limitation on the number of pods per node.
+
+## Design Details
+
+Introducing a new type of NodeAffinity:
+```go
+type NodeSelector struct {
+	NodeSelectorTerms []NodeSelectorTerm
+}
+
+type NodeAffinity struct {
+	RequiredDuringSchedulingRequiredDuringExecution *NodeSelector
+}
+```
+
+Add a controller called `node-affinity-eviction` 
+
+The controller do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the node, checks the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, checks if `NodeSelector` still match the new node. 
+- If `RequiredDuringSchedulingRequiredDuringExecution` is no loger met, trying to evict the pod.
+- It will add a DisruptionTarget condition with corresponding reason and message for the pod.
+- It will use Eviction API with PDB to evict the pod. 
+- If the eviction fails, the pod will remain unchanged with a new pod condition.
+
+New condition:
+```go
+		v1.PodCondition{
+			Type:    v1.DisruptionTarget,
+			Status:  v1.ConditionTrued,
+			Reason:  "DeletionByNodeAffinityEviction",
+			Message: "deleting due to mismatched node affinities",
+		}
+```
```

Comment:
we should include an impact on the daemonset controller. It does its own scheduling and should be tested, that we are not causing a hot loop between scheduling and descheduling.

Reply:
Added to `Integration tests`